src/librustc/infer/region_constraints/mod.rs

```diff
@@ -0,0 +1,909 @@
+
+    /// We added the given `verify`
+    AddVerify(usize),
+
+    /// We added the given `given`
+    AddGiven(Region<'tcx>, ty::RegionVid),
+
+    /// We added a GLB/LUB "combination variable"
+    AddCombination(CombineMapType, TwoRegions<'tcx>),
+
+    /// During skolemization, we sometimes purge entries from the undo
+    /// log in a kind of minisnapshot (unlike other snapshots, this
+    /// purging actually takes place *on success*). In that case, we
+    /// replace the corresponding entry with `Noop` so as to avoid the
+    /// need to do a bunch of swapping. (We can't use `swap_remove` as
+    /// the order of the vector is important.)
+    Purged,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum CombineMapType {
+    Lub,
+    Glb,
+}
+
+type CombineMap<'tcx> = FxHashMap<TwoRegions<'tcx>, RegionVid>;
+
+pub struct RegionSnapshot {
+    length: usize,
+    region_snapshot: unify::Snapshot<ty::RegionVid>,
+    skolemization_count: u32,
+}
+
+/// When working with skolemized regions, we often wish to find all of
+/// the regions that are either reachable from a skolemized region, or
+/// which can reach a skolemized region, or both. We call such regions
+/// *tained* regions.  This struct allows you to decide what set of
+/// tainted regions you want.
+#[derive(Debug)]
+pub struct TaintDirections {
+    incoming: bool,
+    outgoing: bool,
+}
+
+impl TaintDirections {
+    pub fn incoming() -> Self {
+        TaintDirections {
+            incoming: true,
+            outgoing: false,
+        }
+    }
+
+    pub fn outgoing() -> Self {
+        TaintDirections {
+            incoming: false,
+            outgoing: true,
+        }
+    }
+
+    pub fn both() -> Self {
+        TaintDirections {
+            incoming: true,
+            outgoing: true,
+        }
+    }
+}
+
+impl<'tcx> RegionConstraintCollector<'tcx> {
+    pub fn new() -> RegionConstraintCollector<'tcx> {
+        RegionConstraintCollector {
+            var_origins: VarOrigins::default(),
+            data: RegionConstraintData::default(),
+            lubs: FxHashMap(),
+            glbs: FxHashMap(),
+            skolemization_count: 0,
+            bound_count: 0,
+            undo_log: Vec::new(),
+            unification_table: UnificationTable::new(),
+        }
+    }
+
+    pub fn var_origins(&self) -> &VarOrigins {
+        &self.var_origins
+    }
+
+    /// Once all the constraints have been gathered, extract out the final data.
+    ///
+    /// Not legal during a snapshot.
+    pub fn into_origins_and_data(self) -> (VarOrigins, RegionConstraintData<'tcx>) {
+        assert!(!self.in_snapshot());
+        (self.var_origins, self.data)
+    }
+
+    /// Takes (and clears) the current set of constraints. Note that the set of
+    /// variables remains intact.
+    ///
+    /// Not legal during a snapshot.
+    pub fn take_and_reset_data(&mut self) -> RegionConstraintData<'tcx> {
+        mem::replace(&mut self.data, RegionConstraintData::default())
```

Comment:
Shouldn't this also reset the other fields (e.g. `lubs` and `glbs` sound important), to avoid leakage?

Reply:
Hmm. I think you are correct about lubs and glbs. At first I think they were ok, but I guess then we would be caching a LUB from some other locale that may not be a LUB here. Good catch.

Reply:
(To be clear: We're not actually using LUB/GLB in NLL, but it seems good to be thorough anyhow.)

Reply:
Addressed. I also was forgetting about `unification_table`. =)